utils/gcp/object_storage.cc

```diff
@@ -0,0 +1,820 @@
+    _generation = std::stoull(rjson::get<std::string>(item, "generation"));
+}
+
+utils::gcp::storage::client::client(std::string_view endpoint, std::optional<google_credentials> c, shared_ptr<seastar::tls::certificate_credentials> certs)
+    : _impl(seastar::make_shared<impl>(endpoint, std::move(c), std::move(certs)))
+{}
+
+utils::gcp::storage::client::~client() = default;
+
+
+future<> utils::gcp::storage::client::create_bucket(std::string_view project, rjson::value meta) {
+    gcp_storage.debug("Create bucket {}:{}", project, rjson::get(meta, "name"));
+
+    auto path = fmt::format("/storage/v1/b?project={}", project);
+    auto body = rjson::print(meta);
+
+    auto res = co_await _impl->send_with_retry(path
+        , GCP_OBJECT_SCOPE_FULL_CONTROL
+        , body
+        , APPLICATION_JSON
+        , httpclient::method_type::POST
+    );
+
+    switch (res.result()) {
+    case status_type::ok:
+    case status_type::created:
+        co_return; // done and happy
+    default:
+        throw failed_operation(fmt::format("Could not create bucket {}: {}", rjson::get(meta, "name"), res.result()));
+    }
+}
+
+future<> utils::gcp::storage::client::create_bucket(std::string_view project, std::string_view bucket, std::string_view region, std::string_view storage_class) {
+    // Construct metadata. Could fmt::format, but this is somewhat safer.
+    rjson::value meta = rjson::empty_object();
+    rjson::add(meta, "name", std::string(bucket));
+    rjson::add(meta, "location", std::string(region.empty() ? "US" : region));
+    rjson::add(meta, "storageClass", std::string(storage_class.empty() ? "STANDARD" : storage_class));
+
+    rjson::value uniformBucketLevelAccess = rjson::empty_object();
+    rjson::add(uniformBucketLevelAccess, "enabled", true);
+    rjson::value iamConfiguration = rjson::empty_object();
+    rjson::add(iamConfiguration, "uniformBucketLevelAccess", std::move(uniformBucketLevelAccess));
+    rjson::add(meta, "iamConfiguration", std::move(iamConfiguration));
+
+    co_await create_bucket(project, std::move(meta));
+}
+
+future<> utils::gcp::storage::client::delete_bucket(std::string_view bucket) {
+    gcp_storage.debug("Delete bucket {}", bucket);
+
+    auto path = fmt::format("/storage/v1/b/{}", bucket);
+
+    auto res = co_await _impl->send_with_retry(path
+        , GCP_OBJECT_SCOPE_FULL_CONTROL
+        , ""s
+        , ""s
+        , httpclient::method_type::DELETE
+    );
+
+    switch (res.result()) {
+    case status_type::ok: // mock server sends wrong code, but seems acceptable
+    case status_type::no_content:
+        co_return; // done and happy
+    default:
+        throw failed_operation(fmt::format("Could not delete bucket {}: {}", bucket, res.result()));
+    }
+}
+
+// See https://cloud.google.com/storage/docs/listing-buckets
+future<std::vector<utils::gcp::storage::object_info>> utils::gcp::storage::client::list_objects(std::string_view bucket, std::string_view prefix) {
```

Comment:
you have to align here, list what, buckets or objects?

Reply:
Objects. As the method is named. The doc name means "list _contents_ of buckets", but that is just another name for objects.